### ISSUE TITLE
entry processor with member shutdown, throws exception but is executed correctly.

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#12952]
from this run https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-All/60/console
of `shutdown` test, on 3.10

we had multi fails, of the same style.
In the 3 entry processor test failures, each time we see the count on the entry is 1 more that expected, which I think means the entry processore call throw and exception to indicate an error
when the processor was really executed.  In a stable cluster these test pass as expected.

`/disk1/jenkins/workspace/shutdown-All/3.10/2018_04_21-03_17_06/stable/processor/off-load/execute-on-predicate Failed`
```
17:31:17 fail HzMember2HZ _processor_off-load_execute-on-predicate_assertIncs_mapBak1HD hzcmd.map.processor.person.PersonIncAssert threadId=0 global.AssertionException: PersonIdentifiedDataSerializable{} BasePerson{personId=42, age=0, firstName='42', lastName='42', salary=42.0, count=181} count != 180
```
http://54.82.84.143/~jenkins/workspace/shutdown-All/3.10/2018_04_21-03_17_06/stable/processor/off-load/execute-on-predicate

`/disk1/jenkins/workspace/shutdown-All/3.10/2018_04_21-03_17_06/stable/processor/off-load/execute-on Failed`

```
fail HzMember3HZ _processor_off-load_execute-on_assertIncs_mapBak1HD hzcmd.map.processor.person.PersonIncAssert threadId=0 global.AssertionException: PersonSerializable{personId=7800, age=0, firstName='7800', lastName='7800', salary=7800.0, count=1897} count != 1896
``` 
http://54.82.84.143/~jenkins/workspace/shutdown-All/3.10/2018_04_21-03_17_06/stable/processor/off-load/execute-on

`/disk1/jenkins/workspace/shutdown-All/3.10/2018_04_21-03_17_06/stable/processor/regular/execute-on-predicate Failed`

```
fail HzMember1HZ _processor_regular_execute-on-predicate_assertIncs_mapBak1-IDX hzcmd.map.processor.person.PersonIncAssert threadId=0 global.AssertionException: PersonPortable{} BasePerson{personId=19, age=0, firstName='19', lastName='19', salary=19.0, count=98} count != 97
```
http://54.82.84.143/~jenkins/workspace/shutdown-All/3.10/2018_04_21-03_17_06/stable/processor/regular/execute-on-predicate

the test are https://github.com/hazelcast/hzCmd-bench/tree/zeta/lab/hz/stable/processor

**Comment by Danny-Hazelcast** [comment]
entry processor cluster shutdown test's failed on 3.10,  look like the processore was executed but also throw and exception, and the 
incremented count for the entry is 1 more than expected in all cases.

**Comment by Danny-Hazelcast** [comment]
run https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-processor2/1/console
also show same issue.

**Comment by Danny-Hazelcast** [comment]
this issue is reproduced using 3.11-SNAPSHOT 

https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-processor2/4/console

```
fail HzMember1HZ _processor_off-load_execute-on-predicate_assertIncs_mapBak1HD hzcmd.map.processor.person.PersonIncAssert threadId=0 global.AssertionException: PersonExternalizable{} BasePerson{personId=23, age=0, firstName='23', lastName='23', salary=23.0, count=85} count != 84
```

each time the failure is that the count on the entry has been incremented 1 more time than we expected.

we a making controlled member shutdown while the test is running.
as we are using member shutdown and not kill -9, we expect the number of increments completed by the entry processor, to match the number of successful calls to the entry processor.

**Comment by Danny-Hazelcast** [comment]
e.g
/disk1/jenkins/workspace/shutdown-processor2/3.11-SNAPSHOT/2018_05_29-13_09_06/processor/off-load/execute-on-predicate

```
hz-errors | xargs cat | grep 'AssertionException' | sort | uniq
global.AssertionException: PersonExternalizable{} BasePerson{personId=23, age=0, firstName='23', lastName='23', salary=23.0, count=85} count != 84
global.AssertionException: PersonIdentifiedDataSerializable{} BasePerson{personId=37, age=0, firstName='37', lastName='37', salary=37.0, count=142} count != 141
global.AssertionException: PersonIdentifiedDataSerializable{} BasePerson{personId=42, age=0, firstName='42', lastName='42', salary=42.0, count=168} count != 167
global.AssertionException: PersonIdentifiedDataSerializable{} BasePerson{personId=7, age=0, firstName='7', lastName='7', salary=7.0, count=26} count != 25
```

running
 https://github.com/hazelcast/hzCmd-bench/tree/master/lab/hz/stable/processor/off-load/execute-on-predicate

from https://github.com/hazelcast/hzCmd-bench/blob/master/lab/hz/stable/processor/off-load/execute-on-predicate/ops

we running `ExecuteOnPredicatePersonInc` on HD / heap map with and without and Index.

https://github.com/hazelcast/hzCmd-bench/blob/master/src/main/java/hzcmd/map/processor/person/execute/ExecuteOnPredicatePersonInc.java

runs the entry process on a range of keys, and increments a value in the entry.
if successful, we add 1 to the number of successful call's / increments this thread has done

https://github.com/hazelcast/hzCmd-bench/blob/master/src/main/java/hzcmd/map/processor/person/execute/ExecuteOnPredicatePersonInc.java#L21

**Comment by Danny-Hazelcast** [comment]
all though this issue is seen with or with index enabled,
it looks similar to this issue https://github.com/hazelcast/hazelcast-enterprise/issues/1782 ?

**Comment by Danny-Hazelcast** [comment]
I see that the stable cluster version of this these pass.  so i am assuming it is directly related to 
member shutdown.

we also see the kill -9 version of these test fail as validation, however in the kill -9 case this is expected, due to invocation system.

**Comment by Danny-Hazelcast** [comment]
I think this issue is occuring on the shutdown part of the shutdown restart cycle.

I have run this same test in a increasing cluster.  starting off with 1 member and 4 clients.
and while this test is running I add cluster members.

and this test run under the increasing cluster conditions, has not failed.

**Comment by taburet** [comment]
In the end, this issue is a combination of the following two issues:

1. The know issue with client retries for multi-partition invocations and `HazelcastInstanceNotActiveException` (cc @sancar ).

2. The second one is related to https://github.com/hazelcast/hazelcast/issues/11839; as far as I understand the situation, the same operation is executed twice: normally for the first time and for the second time because of the concurrent retry caused by `OnMemberLeftTask`.

Both issues are somewhat out of my expertise area and considered hard to fix (at least by me): the first one requires serious changes to either the client architecture or to the member life cycle (?), the second one leads to questions like "can we really provide the exactly-once in unstable cluster".

**Comment by Danny-Hazelcast** [comment]
we are using controlled member shutdown, so it's not really an unstable cluster

**Comment by taburet** [comment]
@Danny-Hazelcast You are right, but currently, AFAIU, the only difference between hard shutdown and graceful one from the invocations perspective is that the member being shutdown honestly reports that to the cluster, the rest is the same as if it was detected by a failure detector.

**Comment by Holmistr** [comment]
I have created a PRD for this, since it looks like quite a work.
